minicriu-client.c

```diff
@@ -149,40 +486,46 @@ int minicriu_dump(void) {
 		}
 		int r = syscall(SYS_tkill, tid, MC_THREAD_SIG);
 		__atomic_fetch_sub(&mc_futex_checkpoint, 1, __ATOMIC_SEQ_CST);
+		thread_counter++;
 	}
 	closedir(tasksdir);
 
+	mc_thread_counter = thread_counter + 1;
+	debug_log("thread_counter = %d\n", thread_counter);
+
 	uint32_t current_count;
 	while ((current_count = mc_futex_checkpoint) != 0) {
 		syscall(SYS_futex, &mc_futex_checkpoint, FUTEX_WAIT, current_count);
 	}
 
-	struct sigaction acts[SIGRTMAX];
-	struct sigaction new = { .sa_handler = SIG_DFL };
-	for (int i = 1; i < SIGRTMAX; ++i) {
-		if (sigaction(i, &new, &acts[i])) {
-			char msg[256];
-			snprintf(msg, sizeof(msg), "sigaction checkpoint %d: %m", i);
-			fprintf(stderr, "%s\n", msg);
-		}
-	}
+	// Initialize barrier
+	pthread_barrier_init(&mc_thread_barrier, NULL, mc_thread_counter);
 
-	acts[MC_THREAD_SIG] = oldhnd;
+	// Say to other threads that barrier is initialized
+	__atomic_fetch_add(&mc_barrier_initialization, 1, __ATOMIC_SEQ_CST);
```

Comment:
This is not really an elegant solution, you're using two primitives instead of one. If you don't want to count the number of threads ahead, maybe use a semaphore?
Anyway, this piece of code will be eventually rewritten to support more than 64 threads, and catch races between creating new threads and listing them above.

Reply:
You have a point. I'll look into it on my weekend.